tensorboard/data/server/server.rs

```diff
@@ -52,9 +53,27 @@ impl TensorBoardDataProvider for DataProviderHandler {
         _request: Request<data::ListPluginsRequest>,
     ) -> Result<Response<data::ListPluginsResponse>, Status> {
         let mut res: data::ListPluginsResponse = Default::default();
-        res.plugins.push(data::Plugin {
-            name: "scalars".to_string(),
-        });
+        let runs = self.commit.runs.read().unwrap();
+        let mut plugin_names_set: HashSet<String> = HashSet::new();
+        for run_data in runs.values() {
+            let run_data = run_data.read().unwrap();
+            for scalar_value in run_data.scalars.values() {
+                let metadata: &pb::SummaryMetadata = scalar_value.metadata.as_ref();
+                let plugin_data = &metadata.plugin_data.as_ref();
+                match plugin_data {
+                    Some(d) => {
+                        plugin_names_set.insert(d.plugin_name.clone());
+                    }
+                    _ => {
+                        plugin_names_set.insert(String::new());
+                    }
+                }
+            }
+        }
+        // Move out of set into response's list.
```

Comment:
Indeed—these strings are being moved *out* of the set, and ownership of
them given to the vector. You can tell statically that these are being
moved, not copied (no `.clone()`) or borrowed (types are `T`, not `&T`).